lib/web_ui/test/engine/composition_test.dart

```diff
@@ -128,6 +128,30 @@ Future<void> testMain() async {
                 composingBaseOffset: expectedComposingBase,
                 composingExtentOffset: expectedComposingBase + composingText.length));
       });
+
+      test('should return new composition state - compositing from beginning of text', () {
```

Comment:
This looks great! There are some untested code paths in `determineCompositionState` that we should probably add tests for - mainly the null checks and if `composingBase < 0` (will this still occur with this change?)

Reply:
That works for me @Amir-P